### PR TITLE
Port ec2_elb to boto3 - fixes #30229

### LINKED ISSUE #30229: ec2_elb: Failing with boto exception ( 400 - throttling )
##### ISSUE TYPE

bug report

##### COMPONENT NAME

ec2_elb module

##### ANSIBLE VERSION

Ansible 1.7.2
Boto Version 2.32.1

##### OS / ENVIRONMENT

Redhat Enterprise Linux 6.4 ( Ansible Tower v2.0.0)

##### SUMMARY

:**
I have a process "lights on" which turns on ec2 instances, then subsequently adds said instances to their respective load balancers ( defined by elb_shortname ). I get very intermittent results, sometimes the playbook will complete successfully, even consecutively at times. No matter what I do, ( change logic, implement 'pauses' etc, setting and retrieving facts ) I can not get around this AWS throttling message. 

**

##### STEPS TO REPRODUCE

:**

Invoke playbook that turns on ec2_instances, waits and then places machines with defined elb_shortname variable into its respective group load balancer. 

```
- name: starting instance(s)
  when: instance_id is defined and lights_on|default ("false") == "true"
  local_action: ec2
  args:
    region: 'us-west-2'
    instance_ids: "{{ instance_id }}"
    state: 'running'
    wait: 'yes'
    wait_timeout: '300'
  register: ec2

- name: Pausing, trying to avoid AWS throttling
  pause: minutes=10

- name: registering instance to its respective groups ELB
  # instances that do not require ELBs do not need to run this part of the playbook
  when: elb_shortname is defined and lights_on|default ("false") == "true"
  local_action: ec2_elb
  args:
    region: 'us-west-2'
    state: 'present'
    wait: 'yes'
    wait_timeout: '300'
```

**

##### EXPECTED RESULTS

:**

```
TASK: [roles/lights_on | registering instance to its respective groups ELB] *** 
skipping: [tstmaoradbc01 -> 127.0.0.1] 
skipping: [tstmaoradbt01 -> 127.0.0.1] 
skipping: [tstmaoradbt02 -> 127.0.0.1] 
skipping: [tstmaoramqa01 -> 127.0.0.1] 
skipping: [tstmaoramqe01 -> 127.0.0.1] 
skipping: [tstmaoramem01 -> 127.0.0.1] 
skipping: [tstmaoradbu01 -> 127.0.0.1] 
skipping: [tstmaoramem02 -> 127.0.0.1] 
skipping: [tstmaorabix01 -> 127.0.0.1] 
skipping: [tstmaorarel01 -> 127.0.0.1] 
skipping: [tstmaorarex01 -> 127.0.0.1] 
skipping: [tstmaorarex02 -> 127.0.0.1] 
skipping: [tstmaorarex03 -> 127.0.0.1] 
skipping: [tstmaorarex04 -> 127.0.0.1] 
skipping: [tstmaoraetl01 -> 127.0.0.1] 
changed: [tstmaorawsa01 -> 127.0.0.1] 
changed: [tstmaoramax01 -> 127.0.0.1] 
changed: [tstmaorauia02 -> 127.0.0.1] 
changed: [tstmaorauiw01 -> 127.0.0.1] 
changed: [tstmaoramwx01 -> 127.0.0.1] 
changed: [tstmaorawss01 -> 127.0.0.1] 
changed: [tstmaorawsl01 -> 127.0.0.1] 
changed: [tstmaorawss02 -> 127.0.0.1] 
changed: [tstmaorawsa02 -> 127.0.0.1] 
changed: [tstmaorauia01 -> 127.0.0.1] 
changed: [tstmaoramwx02 -> 127.0.0.1]
```

**

##### ACTUAL RESULTS

:**

```
TASK: [roles/lights_on | registering instance to its respective groups ELB] ***
failed: [tstmaorawss01 -> 127.0.0.1] => {"failed": true, "parsed": false}
invalid output was: Traceback (most recent call last):
  File "/Users/ndobbs/.ansible/tmp/ansible-tmp-1412091621.06-77576803493759/ec2_elb", line 1874, in <module>
    main()
  File "/Users/ndobbs/.ansible/tmp/ansible-tmp-1412091621.06-77576803493759/ec2_elb", line 326, in main
    elb_man.register(wait, enable_availability_zone, timeout)
  File "/Users/ndobbs/.ansible/tmp/ansible-tmp-1412091621.06-77576803493759/ec2_elb", line 159, in register
    self._await_elb_instance_state(lb, 'InService', initial_state, timeout)
  File "/Users/ndobbs/.ansible/tmp/ansible-tmp-1412091621.06-77576803493759/ec2_elb", line 196, in _await_elb_instance_state
    instance_state = self._get_instance_health(lb)
  File "/Users/ndobbs/.ansible/tmp/ansible-tmp-1412091621.06-77576803493759/ec2_elb", line 244, in _get_instance_health
    status = lb.get_instance_health([self.instance_id])[0]
  File "/Library/Python/2.7/site-packages/boto/ec2/elb/loadbalancer.py", line 324, in get_instance_health
    return self.connection.describe_instance_health(self.name, instances)
  File "/Library/Python/2.7/site-packages/boto/ec2/elb/__init__.py", line 547, in describe_instance_health
    [('member', InstanceState)])
  File "/Library/Python/2.7/site-packages/boto/connection.py", line 1166, in get_list
    raise self.ResponseError(response.status, response.reason, body)
boto.exception.BotoServerError: BotoServerError: 400 Bad Request
<ErrorResponse xmlns="http://elasticloadbalancing.amazonaws.com/doc/2012-06-01/">
  <Error>
    <Type>Sender</Type>
    <Code>Throttling</Code>
    <Message>Rate exceeded</Message>
  </Error>
  <RequestId>1a08966d-48b8-11e4-8ddc-e3515a48666b</RequestId>
</ErrorResponse>
```

Copied from original issue: ansible/ansible-modules-core#143

### PR DESCRIPTION
**Pull request by s-hertel**
##### SUMMARY
Fixes #30229 by adding AWSRetry.

##### ISSUE TYPE
 - Bugfix Pull Request

##### COMPONENT NAME
lib/ansible/modules/cloud/amazon/ec2_elb.py

##### ANSIBLE VERSION
```
2.5.0
```

### CODE REVIEW

**Comment by s-hertel** [comment]
Need to add permissions to Shippable and update the policy in hacking/aws_config.

File: lib/ansible/modules/cloud/amazon/ec2_elb.py [hunk 1]
```diff
@@ -353,25 +377,39 @@ def main():
...
+    try:
+        client = boto3_conn(module, conn_type='client', resource='elb', region=region, **aws_connect_params)
+    except (botocore.exceptions.ProfileNotFound, botocore.exceptions.PartialCredentialsError) as e:
+        module.fail_json_aws(e)
+
     elb_man = ElbManager(module, instance_id, ec2_elbs, region=region, **aws_connect_params)
+    try:
+        elb_man.lbs = elb_man._get_instance_lbs(client, ec2_elbs)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e)
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_elb.py]: Please also provide a message here.

File: lib/ansible/modules/cloud/amazon/ec2_elb.py [hunk 2]
```diff
@@ -295,34 +319,37 @@ def _get_auto_scaling_group_lbs(self):
...
     def _get_instance(self):
         """Returns a boto.ec2.InstanceObject for self.instance_id"""
         try:
-            ec2 = connect_to_aws(boto.ec2, self.region, **self.aws_connect_params)
-        except (boto.exception.NoAuthHandlerFound, AnsibleAWSError) as e:
-            self.module.fail_json(msg=str(e))
-        return ec2.get_only_instances(instance_ids=[self.instance_id])[0]
+            ec2 = boto3_conn(self.module, conn_type='client', resource='ec2', region=self.region, **self.aws_connect_params)
+        except botocore.exceptions.ProfileNotFound as e:
+            self.module.fail_json_aws(e)
```
Comment (ryansb) [inline lib/ansible/modules/cloud/amazon/ec2_elb.py]: Message on fail please :smiley:

File: test/integration/targets/ec2_elb/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,536 @@
+---
+- block:
```
Comment (ryansb) [inline test/integration/targets/ec2_elb/tasks/main.yml]: I think `block` has a `environment:` attribute that you can use to set all the aws secret/key id/session token variables as their bash appropriate versions (AWS_SECRET_KEY etc) and not have to do so on every module call. 

For the tests checking that the module fails correctly when missing credentials, put them in a separate block.

File: test/integration/targets/ec2_elb/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,536 @@
...
+        state: present
+        ec2_elbs: "{{ lb_name }}"
+        instance_id: "{{ setup_instance.instance_ids[0] }}"
+      register: result
+      ignore_errors: true
+
+    - name: assert task failed
+      assert:
+        that:
+          - "result.failed"
```
Comment (ryansb) [inline test/integration/targets/ec2_elb/tasks/main.yml]: The preferred way to check failure is to use the filter, so there's no error if that attribute is undefined. Just `result | failed`

**Review by ryansb — CHANGES_REQUESTED** [review]

**Comment by ElijahLynn** [comment]
I have been testing this for some provisioning over the past week and it has saved the day in terms of preventing API throttled errors/messages from AWS!!! So, :+1: for some basic testing so far.

**Comment by wimnat** [comment]
I thought the plan for this was to just leave this module 'as-is' and instead create elb_classic_lb as the boto3 version allowing us to fix return values etc to be more in line with elb_application_lb?

**Comment by s-hertel** [comment]
@wimnat Ah....Sigh. Yes, that was the plan. Thanks for noticing this!

**Comment by s-hertel** [comment]
@wimnat Are you working on elb_classic_lb for 2.5?

**Comment by wimnat** [comment]
I'd like to but not started yet. However my head is in the space now - just pushed WIP PR for elb_network_lb.

**Comment by ElijahLynn** [comment]
Here is the WIP PR referenced by @wimnat above ==> https://github.com/ansible/ansible/pull/33808

**Comment by jchristi** [comment]
The above linked PR is for [elb_network_lb](https://github.com/wimnat/ansible/blob/3150abae56f304f6091ae7d1eb42f782de1da501/lib/ansible/modules/cloud/amazon/elb_network_lb.py), not [elb_classic_lb](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/elb_classic_lb.py) or [ec2_elb_lb](https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/cloud/amazon/ec2_elb_lb.py). A quick look at those two modules show they are both still using old boto. Is there anywhere boto3 conversion for a classic elb is being tracked? @ElijahLynn and I really want this!

**Comment by ElijahLynn** [comment]
@s-hertel Do you know if this functionality is somehow in the recent 2.6 release? I see https://github.com/ansible/ansible/blame/devel/lib/ansible/modules/cloud/amazon/elb_application_lb.py#L28 uses boto3 but did not see any reference of exponential or backoff, but that doesn't mean for sure it isn't included.

**Comment by s-hertel** [comment]
@wimnat @ElijahLynn @jchristi Hi, all. Linking a pull request to port elb_classic_lb to boto3 for 2.8. If any of you will use it, testing/reviewing what is there would be greatly appreciated. https://github.com/ansible/ansible/pull/45434

@ElijahLynn Are you having throttling issues with the elb_application_lb module? elb_application_lb is boto3 and uses AWSRetry (the retried calls are being done in module_utils/aws/elbv2.py and module_utils/aws/elb_utils.py).

**Comment by jchristi** [comment]
@s-hertel I have been using elb_application_lb and have yet to see any API throttling there. I doubt I will be able to find the time to test classic elb, but I'll let you know if I do.

**Comment by jamiecwilliams** [comment]
> I thought the plan for this was to just leave this module 'as-is' and instead create elb_classic_lb as the boto3 version allowing us to fix return values etc to be more in line with elb_application_lb?

> @wimnat Ah....Sigh. Yes, that was the plan. Thanks for noticing this!

@s-hertel, isn't this PR for porting the `ec2_elb` module? The `ec2_classic_lb` module should not be relevant, because it is for managing the lbs, where `ec2_elb` is for registering instances. As far as I can tell, there is no alternative to `ec2_elb` which is not boto3 based, and it is still lacking AWSRetry support. @ryansb added `ec2_elb` to the list of "boto only" modules very recently: https://github.com/ansible/community/blob/master/group-aws/boto3.md

Perhaps @wimnat confused your effort with decisions over the `ec2_elb_lb` module, which has been superceded (if I understand correctly) by `ec2_classic_lb`, `ec2_network_lb`, and `ec2_appliance_lb`?
Issue https://github.com/ansible/ansible/issues/30494 regarding renaming or replacing existing lb-related modules, suggested `ec2_instance` as a new boto3 module. However, `ec2_instance` is also in the list of "boto only" modules. The current documentation does not indicate any retry capability in this new module either.

I'm having major difficulty due to throttling with the `ec2_elb` module. Is there any reason `elb_instance` would reduce throttling problems? What is the difference between `ec2_elb` and `ec2_instance` if they both use boto? Is there still a plan to update either to boto3?

Since issue https://github.com/ansible/ansible/issues/30229 references this PR and is still open, and this is still a current problem with this module, I request that this PR be re-opened or similar changes be made to the `ec2_instance` module.

**Pull request closed without merging**